### PR TITLE
CDAP-19256 : Conversion of KeyValue Widget testcases to selenium

### PR DESCRIPTION
**Pull request by radhikav1**
# Convert widgets.KeyValueWidget.spec.ts test from cypress to selenium

## Description
Conversion of KeyValue Widget testcases to selenium

## PR Type
- [ ] Bug Fix
- [ ] Feature
- [ ] Build Fix
- [x] Testing
- [ ] General Improvement
- [ ] Cherry Pick

## Links
Jira: [CDAP-19256](https://cdap.atlassian.net/browse/CDAP-19256)

## Test Plan
Included in e2e tests

## Screenshots
n/a

[CDAP-19256]: https://cdap.atlassian.net/browse/CDAP-19256?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/e2e-test/features/widget.csv.feature [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      When Open Pipeline Studio Page
+      Then Toggle Transform Panel
+      Then Add Projection node to canvas
+      Then Open Projection node properties
+      Then Click add row and enter "value1", "value2" for row 1 and row 2 respectively
+      Then Delete row 2
+      Then Close Projection node properties
+      Then Open Projection node properties
+      Then Verify CSV Widget values "value1", "value2" in Pipeline Stage JSON
+      Then Exit Studio Page
```
Comment (njbriggs) [inline src/e2e-test/features/widget.csv.feature]: Nit: newline at end of file

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java [hunk 2]
```diff
@@ -299,6 +299,12 @@ public static void setShowWelcomeSessionStorage() {
       "window.sessionStorage.setItem('showWelcome','true');"));
   }
 
+  public static void setWindowPropertyCypress() {
+    JavascriptExecutor js = (JavascriptExecutor) SeleniumDriver.getDriver();
+    js.executeScript(String.format(
+            "window.Cypress = true"));
```
Comment (njbriggs) [inline src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: In some places we look for `window.Cypress.cy.on`. We should probably set `window.Cypress` to be an object with a no-op `cy.on` function.

File: src/e2e-test/features/widget.csv.feature (hunk 1 above)
Comment (radhikav1) [reply src/e2e-test/features/widget.csv.feature]: done

File: src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java (hunk 2 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/utils/Helper.java]: For now I've added Cypress object with `on` function

File: src/e2e-test/features/widget.keyvalue.feature [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  Scenario: Should delete property in KeyValue Widget
+    When Open Pipeline Studio Page
+    Then Toggle Transform Panel
+    Then Add Projection node to canvas
+    Then Open Projection node properties
+    Then Add key value pairs as "key1", "value1" and "key2", "value2" respectively
+    Then Delete keyValue row 2 and verify
+    Then Close Projection node properties
+    Then Verify key value pairs "key1:value1" in Pipeline Stage JSON
+    Then Exit Studio Page
```
Comment (seanfeld) [inline src/e2e-test/features/widget.keyvalue.feature]: new line

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetCSV.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
```
Comment (seanfeld) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetCSV.java]: were these files made in 2022 or 2023? some files below have 2023 - lets be consistent and pick one for this pr

Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetCSV.java]: As the WidgetCSV tests are merged now, this PR holds only required files.

File: src/e2e-test/features/widget.keyvalue.feature (hunk 3 above)
Comment (radhikav1) [reply src/e2e-test/features/widget.keyvalue.feature]: done

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetKeyValue.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+
+public class WidgetKeyValue {
+
+  private String getCssSelectorForRow(String rowIndex) {
+    return Helper.getCssSelectorByDataTestId("rename") + " " + Helper.getCssSelectorByDataTestId(rowIndex);
+  }
+
+  private String getCssSelectorForKey(String rowIndex) {
```
Comment (sumengwang) [inline src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetKeyValue.java]: seems like this is a duplicated function that exists in your another pr?

**Review by sumengwang — CHANGES_REQUESTED** [review]
looks like you have many overlapping functions that are in https://github.com/cdapio/cdap-ui/pull/888
please refactor to a more generic usage

File: src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetKeyValue.java (hunk 5 above)
Comment (radhikav1) [reply src/e2e-test/java/io/cdap/cdap/ui/stepsdesign/WidgetKeyValue.java]: Made the these functions generic as possible and moved to Helper class to use in other Widget testcases too.

**Comment by radhikav1** [comment]
> looks like you have many overlapping functions that are in #888 please refactor to a more generic usage

I've made functions as generic and moved to Helper class. Please resume the review

**Review by sumengwang — APPROVED** [review]

**Review by seanfeld — APPROVED** [review]

**Pull request merged**